pkg/local/diff.go

```diff
@@ -188,6 +186,59 @@ func Undo(c Changes) Changes {
 	return reversed
 }
 
+// WalkS3 - walk like an Egyptian... ¯\_(ツ)¯\_
+// This walker function simulates the way object listing is performed by S3. Contrary to how a standard FS walk function behaves, S3
+// does not take into consideration the directory hierarchy. Instead, object paths include the entire path relative to the root and as a result
+// the directory or "path separator" is also taken into account when providing the listing in a lexicographical order.
+func WalkS3(root string, callbackFunc func(p string, info fs.FileInfo, err error) error) error {
+	var dirQueue []string
+	err := filepath.Walk(root, func(p string, info fs.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if p == root {
+			return nil
+		}
+		if info.IsDir() {
+			// TODO: We don't return dir results for the listing, how will this effect directory markers, and can we even support directory markers?
+			// Save encountered directories in a queue and compare them with the first appearance of an file in that level
+			dirQueue = append(dirQueue, p+path.Separator) // add path separator to dir name and sort it later
+			return filepath.SkipDir
+		}
+
+		// Sort the queue since adding the suffix might the order might change after adding the suffix
+		sort.Strings(dirQueue)
+		// It is enough to compare the file with the last encountered directory (suffixed by the PathSeparator)
+		// if file > last_dir + "/" then by transitiveness it is also > than all previous directories
+		queueLen := len(dirQueue)
+		if queueLen > 0 && p > dirQueue[len(dirQueue)-1] {
+			// if file > dirs, handle the dirs first
+			for _, dir := range dirQueue {
+				if err = WalkS3(dir, callbackFunc); err != nil {
+					return err
+				}
+			}
+			dirQueue = []string{} // Empty queue once finished processing dirs
+		}
```

Comment:
I do not think that this can be correct.  Consider the case
```go
dirQueue[0] < p && p < dirQueue[len(dirQueue)-1]
```

Reply:
That is impossible:
for simplicity lets call dirQueue[len(dirQueue)-1] `last`
and dirQueue[0] `first`
if p < `last` ==> p < `last - "/"` (because otherwise the walk function would have returned p before `last`
if `first` < p ==> `first - "/"` < p
Therefore it mandates that  `first - "/"` < p < `last - "/"` which means that we should have gotten p from the walk function somewhere between getting `first` and `last` and we get a contradiction

Reply:
I hope the explanation is convincing 😃

Reply:
I'm retracting my previous statement, I have not taken into account different string length. I have added the code that deals with this scenario as well as a unit test